File: uspec/templates/capture.js
// Capture the uSpec template components into a portable JSON description.
//
// Run through the bridge against the file that holds the templates:
//   python capture.py --session "uSpec template (Community)"
//
// Read-only. Nothing is created, moved or modified.
//
// Why a capture rather than a hand-written builder: AUTHORING these templates in JS is a
// trap — 147 named anchor layers whose geometry every
// render script assumes, drifting from upstream on every `uspec-skills update`. Capture-and-
// replay is a different thing: the description is read off the real component by machine, and
// replay produces real components with real #anchor layers, so the render scripts and their
// upgrades are unaffected.
//
// The templates were measured first (see ../uspec-templates.md): 330 nodes, only COMPONENT /
// FRAME / TEXT / VECTOR, and zero bound variables, styles, images, instances, effects or
// component properties. That is what makes a faithful capture possible at all — there is
// nothing in them that belongs to the file they live in. This script asserts those invariants
// and fails loudly if a future template version breaks them.

await figma.loadAllPagesAsync();

const WANT = ["Screen reader", "Color Annotation", "Anatomy", "API", "Property", "Structure", "Motion"];

const tops = figma.root
  .findAllWithCriteria({ types: ["COMPONENT"] })
  .filter((n) => !(n.parent && n.parent.type === "COMPONENT_SET"));

const missing = WANT.filter((w) => !tops.some((t) => t.name === w));
if (missing.length) throw new Error("templates not found in " + figma.root.name + ": " + missing.join(", "));

const violations = [];
const fonts = {};

// Properties worth carrying. Anything absent on a node is simply skipped, so one list serves
// all four node types.
const SCALARS = [
  "visible", "opacity", "blendMode", "isMask", "clipsContent",
  "layoutMode", "layoutWrap", "layoutAlign", "layoutGrow", "layoutPositioning",
  "primaryAxisSizingMode", "counterAxisSizingMode",
  "primaryAxisAlignItems", "counterAxisAlignItems",
  "layoutSizingHorizontal", "layoutSizingVertical",
  "itemSpacing", "counterAxisSpacing", "itemReverseZIndex",
  "paddingTop", "paddingRight", "paddingBottom", "paddingLeft",
  "topLeftRadius", "topRightRadius", "bottomLeftRadius", "bottomRightRadius",
  "strokeWeight", "strokeAlign", "strokeCap", "strokeJoin", "strokeMiterLimit", "dashPattern",
  // Per-side weights: three frames (#ruler-track, #tick, #track-area) draw a rule on one edge
  // only, which reads as `strokeWeight: mixed`. Capturing the scalar alone would replay them
  // as a full box.
  "strokeTopWeight", "strokeRightWeight", "strokeBottomWeight", "strokeLeftWeight",
  "rotation", "cornerSmoothing", "overflowDirection",
  // TEXT
  "characters", "fontSize", "textAlignHorizontal", "textAlignVertical", "textAutoResize",
  "textCase", "textDecoration", "paragraphSpacing", "paragraphIndent", "textTruncation",
  "maxLines", "autoRename", "hangingPunctuation", "hangingList", "leadingTrim",
  // VECTOR
  "handleMirroring",
];

const val = (v) => (v === figma.mixed ? "__MIXED__" : v);

function capture(n) {
  // Invariants: anything below would not survive a move to another file.
  if (n.boundVariables && Object.keys(n.boundVariables).length)
    violations.push(n.name + ": boundVariables");
  for (const p of ["fillStyleId", "strokeStyleId", "textStyleId", "effectStyleId", "gridStyleId"])
    if (typeof n[p] === "string" && n[p]) violations.push(n.name + ": " + p);
  if (n.type === "INSTANCE") violations.push(n.name + ": nested INSTANCE");
  if (Array.isArray(n.fills))
    for (const f of n.fills)
      if (f.type === "IMAGE" || f.type === "VIDEO") violations.push(n.name + ": " + f.type + " fill");

  const out = { name: n.name, type: n.type, width: n.width, height: n.height };

  for (const p of SCALARS) {
    const v = n[p];
    if (v !== undefined && v !== null) out[p] = val(v);
  }

  // A child of an auto-layout parent gets its position from the layout; a child of a plain
  // frame needs x/y. Carry both — replay decides which to apply.
  if (typeof n.x === "number") { out.x = n.x; out.y = n.y; }
  if (n.constraints) out.constraints = n.constraints;

  if (Array.isArray(n.fills)) out.fills = JSON.parse(JSON.stringify(n.fills));
  if (Array.isArray(n.strokes)) out.strokes = JSON.parse(JSON.stringify(n.strokes));
  if (Array.isArray(n.effects) && n.effects.length) out.effects = JSON.parse(JSON.stringify(n.effects));

  if (n.type === "TEXT") {
    const f = n.fontName;
    if (f === figma.mixed) {
      // Mixed runs: capture per-character-range so replay can rebuild them.
      out.fontName = "__MIXED__";
      out.runs = [];
      let i = 0;
      while (i < n.characters.length) {
        const end = n.getRangeAllFontNames(i, i + 1) && i + 1;
        const fn = n.getRangeFontName(i, i + 1);
        const last = out.runs[out.runs.length - 1];
        if (last && last.family === fn.family && last.style === fn.style) last.end = end;
        else out.runs.push({ start: i, end: end, family: fn.family, style: fn.style });
        i++;
      }
      for (const r of out.runs) fonts[r.family + "|" + r.style] = true;
    } else {
      out.fontName = { family: f.family, style: f.style };
      fonts[f.family + "|" + f.style] = true;
    }
    out.lineHeight = val(n.lineHeight);
    out.letterSpacing = val(n.letterSpacing);
  }

  if (n.type === "VECTOR") {
    out.vectorPaths = JSON.parse(JSON.stringify(n.vectorPaths));
  }

  if (n.children) out.children = n.children.map(capture);
  return out;
}

const templates = {};
for (const w of WANT) templates[w] = capture(tops.find((t) => t.name === w));

// Anything still reading `mixed` after the per-side weights are captured is a real fidelity
// gap. Surfaced rather than swallowed, so replay never quietly draws something else.
const gaps = [];
const SIDE_WEIGHTS = ["strokeTopWeight", "strokeRightWeight", "strokeBottomWeight", "strokeLeftWeight"];
const scanGaps = (o, path) => {
  for (const [k, v] of Object.entries(o)) {
    if (v !== "__MIXED__" || k === "fontName") continue;
    // A mixed strokeWeight is fully described by the four per-side weights, so it replays
    // exactly. Not a gap.
    if (k === "strokeWeight" && SIDE_WEIGHTS.every((s) => typeof o[s] === "number")) continue;
    gaps.push({ node: path + "/" + o.name, type: o.type, property: k });
  }
  for (const c of o.children || []) scanGaps(c, path + "/" + o.name);
};

for (const w of WANT) scanGaps(templates[w], w);

const countNodes = (o) => 1 + (o.children || []).reduce((n, c) => n + countNodes(c), 0);
const countAnchors = (o) =>
  (o.name.charAt(0) === "#" ? 1 : 0) + (o.children || []).reduce((n, c) => n + countAnchors(c), 0);

return {
  capturedFrom: figma.root.name,
  capturedAt: new Date().toISOString(),
  fontsRequired: Object.keys(fonts).map((k) => ({ family: k.split("|")[0], style: k.split("|")[1] })),
  violations: violations,
  fidelityGaps: gaps,
  perTemplate: WANT.map((w) => ({
    name: w,
    nodes: countNodes(templates[w]),
    anchors: countAnchors(templates[w]),
  })),
  totalNodes: WANT.reduce((n, w) => n + countNodes(templates[w]), 0),
  totalAnchors: WANT.reduce((n, w) => n + countAnchors(templates[w]), 0),
  templates: templates,
};

File: uspec/templates/capture.py
#!/usr/bin/env python3
"""Capture the uSpec templates out of Figma into a portable JSON description.

    python capture.py --session "uSpec template (Community)"

Read-only. Posts capture.js through the Figmosha bridge and writes templates.json next to it.
The description is what replay.py rebuilds from, and what git diffs when the templates change
upstream.
"""

from __future__ import annotations

import argparse
import json
import sys
import time
import urllib.error
import urllib.request
from pathlib import Path

def default_port() -> int:
    """This copy's port, from project.json — the same number the bridge binds."""
    try:
        sys.path.insert(0, str(Path(__file__).resolve().parents[2]))
        import project
        cfg = project.load()
        if cfg:
            return int(cfg["port"])
    except Exception:
        pass
    return 8787


HERE = Path(__file__).resolve().parent


def main() -> None:
    p = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    p.add_argument("--session", help="Figma file holding the templates, when several are connected")
    p.add_argument("--port", type=int, default=default_port(),
                   help="bridge port (default: this copy's project.json)")
    p.add_argument("--timeout", type=int, default=240)
    p.add_argument("--out", default=str(HERE / "templates.json"))
    args = p.parse_args()

    payload = {
        "code": (HERE / "capture.js").read_text(encoding="utf-8"),
        "timeout": args.timeout,
        "want_value": True,
    }
    if args.session:
        payload["session"] = args.session

    req = urllib.request.Request(
        "http://localhost:%d/exec" % args.port,
        data=json.dumps(payload, ensure_ascii=True).encode("utf-8"),
        headers={"Content-Type": "application/json"},
    )
    t0 = time.time()
    try:
        with urllib.request.urlopen(req, timeout=args.timeout + 15) as r:
            body = r.read()
    except urllib.error.HTTPError as e:
        body = e.read()
    except urllib.error.URLError as e:
        sys.exit("bridge unreachable on port %d (%s)" % (args.port, e))

    resp = json.loads(body)
    if not resp.get("ok"):
        sys.stderr.write("capture failed: %s\n%s\n" % (resp.get("error"), (resp.get("stack") or "")[:1500]))
        sys.exit(1)

    v = resp["value"]

    if v["violations"]:
        sys.stderr.write(
            "\nREFUSING TO CAPTURE — the templates now contain things that do not survive a move\n"
            "to another file. Replay would silently produce a different frame:\n"
        )
        for x in v["violations"]:
            sys.stderr.write("  - %s\n" % x)
        sys.stderr.write(
            "\nThis check is the whole reason capture-and-replay is safe. Do not remove it;\n"
            "re-read ../uspec-templates.md and decide what the new content needs.\n"
        )
        sys.exit(1)

    Path(args.out).write_text(json.dumps(v, ensure_ascii=True, indent=2), encoding="utf-8")

    print("captured from %s in %.1f s" % (v["capturedFrom"], time.time() - t0))
    for t in v["perTemplate"]:
        print("  %-18s %3d nodes, %3d anchors" % (t["name"], t["nodes"], t["anchors"]))
    print("  %-18s %3d nodes, %3d anchors" % ("TOTAL", v["totalNodes"], v["totalAnchors"]))
    print("  fonts required: %s" % ", ".join("%s %s" % (f["family"], f["style"]) for f in v["fontsRequired"]))
    print("  -> %s (%.0f KB)" % (args.out, Path(args.out).stat().st_size / 1024))

    gaps = v.get("fidelityGaps") or []
    if gaps:
        print("\n  %d fidelity gap(s) — properties that read `mixed` and replay cannot reproduce" % len(gaps))
        for g in gaps:
            print("    %-10s %-16s %s" % (g["type"], g["property"], g["node"]))
        print("  These are recorded, not silent. Check them against the replayed frame.")


if __name__ == "__main__":
    main()
